File: codes/code_feb_2024/linked_list/linked_list.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __str__(self) -> str:
        ret = str(self.val)
        if self.next:
            ret += self.next.__str__()
        return ret


class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None
        front_node: ListNode = head
        cur = front_node
        count = 0
        while cur.next:
            cur = cur.next
            count += 1
        if count < 2:
            return front_node
        back_node = cur
        back_cur = back_node
        cur = front_node
        count = (count + 1) // 2
        while count > 0:
            if cur and cur.next:
                node: ListNode = cur.next
                cur.next = cur.next.next
                back_cur.next = ListNode(node.val)
                back_cur = back_cur.next
                cur = cur.next
            else:
                break
            count -= 1
        return front_node


if __name__ == "__main__":
    s = Solution()
    lst = list(range(1, 4))
    front = ListNode()
    cur_temp = front
    for num in lst:
        cur_temp.next = ListNode()
        cur_temp.next.val = num
        cur_temp = cur_temp.next
    print(front)
    print(s.oddEvenList(front.next))

File: codes/code_feb_2024/linked_list/linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
#include <iostream>
#include <memory>
#include <string>
#include <vector>
using std::cout;
using std::endl;
using std::string;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
    void print()
    {
        ListNode *node = this;
        while (node)
        {
            cout << node->val;
            if (!node->next)
            {
                cout << endl;
            }
            else
            {
                cout << " ";
            }
            node = node->next;
        }
    }
};

ListNode *push(ListNode *node, int val)
{
    ListNode *nextNode = new ListNode(val);
    node->next = nextNode;
    return node->next;
}

class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        ListNode *pre_node = nullptr;
        ListNode *next_node = nullptr;
        ListNode *cur = head;
        while (cur)
        {
            cout << cur->val << endl;
            next_node = cur->next;
            cur->next = pre_node;
            pre_node = cur;
            cur = next_node;
        }
        return pre_node;
    }

    int pairSum(ListNode *head)
    {
        std::vector<int> vec = {};
        ListNode *cur = head;
        while (cur) {
            vec.emplace_back(cur->val);
            cur = cur->next;
        }
        int max_sofar = 0;
        int length = vec.size();
        if (length == 0) {
            return max_sofar;
        }
        int temp = 0;
        for (int i = 0; i < length; ++i) {
            temp = vec[i] + vec[length - 1 - i];
            if (temp > max_sofar) {
                max_sofar = temp;
            }
        }
        return max_sofar;
    }
};

int main(int argc, char const *argv[])
{
    int arr[5] = {1, 2, 3, 4, 5};
    ListNode fakehead = ListNode(0);
    ListNode *cur = &fakehead;
    for (auto val : arr)
    {
        cur = push(cur, val);
    }
    fakehead.print();
    ::Solution s = Solution();
    fakehead.next = s.reverseList(fakehead.next);
    fakehead.next->print();
    return 0;
}
